src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,178 @@
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (size == 0) {
+            Node<T> first = new Node<>(null, value, null);
+            head = first;
+            tail = first;
+            first.prev = null;
+            first.next = null;
+            size++;
+        } else if (index == size) {
+            Node<T> last = tail;
+            Node<T> newNode = new Node<>(last, value, null);
+            newNode.prev = tail;
+            last.next = newNode;
+            tail = newNode;
+            size++;
+        } else {
+            Node<T> current = node(index);
+            Node<T> left = current.prev;
+            Node<T> newNode = new Node<>(left, value, current);
+            if (left == null) {
+                head = newNode;
+            } else {
+                current.prev = newNode;
+                left.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexChecker(index);
+        return node(index).items;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexChecker(index);
+        Node<T> setElement = head;
+        for (int i = 0; i < index; i++) {
+            setElement = head.next;
+        }
+        T oldValue = setElement.items;
+        setElement.items = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexChecker(index);
+        Node<T> oldNode = node(index);
+        Node<T> prevNode = node(index).prev;
+        Node<T> nextNode = node(index).next;
+        final T returnElement = oldNode.items;
+        if (oldNode.prev == null) {
+            head = nextNode;
+        } else {
+            prevNode.next = nextNode;
+            oldNode.prev = null;
+        }
+        if (oldNode.next == null) {
+            tail = prevNode;
+        } else {
+            nextNode.prev = prevNode;
+            oldNode.next = null;
+        }
+        oldNode.items = null;
+        size--;
+        return returnElement;
     }
 
     @Override
     public boolean remove(T object) {
+        int counter = 0;
+        Node<T> header = head;
+        if (object == null) {
+            for (int i = 0; i < size; i++) {
+                if (header.items == null) {
+                    remove(i);
+                    return true;
+                }
+                header = head.next;
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (header.items.equals(object)) {
+                    remove(i);
+                    return true;
+                }
+                header = head.next;
+            }
```

Comment:
what's the point of iterating twice? you can compare two objects and avoid NPE at the same time